services/shhext/chat/encryption.go

```diff
@@ -0,0 +1,468 @@
+	bundleContainer, err := s.persistence.GetAnyPrivateBundle(ourIdentityKeyC)
+	if err != nil {
+		return nil, err
+	}
+
+	// If the bundle has expired we create a new one
+	if bundleContainer != nil && bundleContainer.Timestamp < time.Now().AddDate(0, 0, -14).UnixNano() {
+		// Mark sessions has expired
+		if err := s.persistence.MarkBundleExpired(bundleContainer.GetBundle().GetIdentity()); err != nil {
+			return nil, err
+		}
+
+	} else if bundleContainer != nil {
+		err = SignBundle(privateKey, bundleContainer)
+		if err != nil {
+			return nil, err
+		}
+		return bundleContainer.GetBundle(), nil
+	}
+
+	// needs transaction/mutex to avoid creating multiple bundles
+	// although not a problem
+	bundleContainer, err = NewBundleContainer(privateKey, s.installationID)
+	if err != nil {
+		return nil, err
+	}
+
+	if err = s.persistence.AddPrivateBundle(bundleContainer); err != nil {
+		return nil, err
+	}
+
+	return s.CreateBundle(privateKey)
+}
+
+// DecryptWithDH decrypts message sent with a DH key exchange, and throws away the key after decryption
+func (s *EncryptionService) DecryptWithDH(myIdentityKey *ecdsa.PrivateKey, theirEphemeralKey *ecdsa.PublicKey, payload []byte) ([]byte, error) {
+	key, err := PerformDH(
+		ecies.ImportECDSA(myIdentityKey),
+		ecies.ImportECDSAPublic(theirEphemeralKey),
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return crypto.DecryptSymmetric(key, payload)
+
+}
+
+// keyFromPassiveX3DH decrypts message sent with a X3DH key exchange, storing the key for future exchanges
+func (s *EncryptionService) keyFromPassiveX3DH(myIdentityKey *ecdsa.PrivateKey, theirIdentityKey *ecdsa.PublicKey, theirEphemeralKey *ecdsa.PublicKey, ourBundleID []byte) ([]byte, error) {
+	bundlePrivateKey, err := s.persistence.GetPrivateKeyBundle(ourBundleID)
+	if err != nil {
+		s.log.Error("Could not get private bundle", "err", err)
+		return nil, err
+	}
+
+	if bundlePrivateKey == nil {
+		return nil, errSessionNotFound
+	}
+
+	signedPreKey, err := ecrypto.ToECDSA(bundlePrivateKey)
+	if err != nil {
+		s.log.Error("Could not convert to ecdsa", "err", err)
+		return nil, err
+	}
+
+	key, err := PerformPassiveX3DH(
+		theirIdentityKey,
+		signedPreKey,
+		theirEphemeralKey,
+		myIdentityKey,
+	)
+	if err != nil {
+		s.log.Error("Could not perform passive x3dh", "err", err)
+		return nil, err
+	}
+	return key, nil
+}
+
+// ProcessPublicBundle persists a bundle
+func (s *EncryptionService) ProcessPublicBundle(myIdentityKey *ecdsa.PrivateKey, b *Bundle) error {
+	// Make sure the bundle belongs to who signed it
+	err := VerifyBundle(b)
+	if err != nil {
+		return err
+	}
+	return s.persistence.AddPublicBundle(b)
+}
+
+// DecryptPayload decrypts the payload of a DirectMessageProtocol, given an identity private key and the sender's public key
+func (s *EncryptionService) DecryptPayload(myIdentityKey *ecdsa.PrivateKey, theirIdentityKey *ecdsa.PublicKey, msgs *map[string]*DirectMessageProtocol) ([]byte, error) {
```

Comment:
`msgs map[string]*DirectMessageProtocol` will be fine. `map`s are passed by a pointer.